Added isSerializable and SERIALIZABLE to Field

BeksOmega opened the pull request:
<!--
  - Thanks for submitting code to Blockly!  Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## The basics

<!-- TODO: Verify the following, checking each box with an 'x' between the brackets: [x] -->

- [X] I branched from develop
- [X] My pull request is against develop
- [X] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

## The details
### Resolves

<!-- TODO: What Github issue does this resolve? Please include a link. -->
Work on #1623 

### Proposed Changes

<!-- TODO: Describe what this Pull Request does.  Include screenshots if applicable. -->
Adds a SERIALIZABLE property to fields. Adds an isSerializale function that handles warning & backwards compatibility logic.

I also reorganized the test-blocks toolbox so now it looks like this:
![PlaygroundToolboxNew](https://user-images.githubusercontent.com/25440652/56371374-dc060b00-61b1-11e9-8544-92a738e2896d.jpg)

### Reason for Changes

<!--TODO: Explain why these changes should be made.  Include screenshots if applicable. -->
Sometimes you would want to programmatically modify a field, but not allow users to modify it, and save it to XML. Before you would have to create a mutator for this purpose. Now you can simply use a serializable field.

### Test Coverage

<!-- TODO: Please show how you have added tests to cover your changes,
  -        or tell us how you tested it. For each systems you tested,
  -        uncomment the systems in the list below.
  -->

I added unit tests! I added tests for the isSerializable function to make sure it works in a backwards compatible way. I also added serialization tests to make sure the XML for different fields was comming out correctly.

Here's also a quick example of some fields being serialized and deserialized:
![Serialization](https://user-images.githubusercontent.com/25440652/56371892-d3fa9b00-61b2-11e9-96c6-9087a2f29fb7.gif)

Here's the generated XML for that example:
```
<xml xmlns="http://www.w3.org/1999/xhtml">
  <variables>
    <variable type="" id="o(EVLYWaZFZca|N:K.7I">not item</variable>
  </variables>
  <block type="test_fields_angle" id="mkWW9%.I_@;R#$W[H{PM" x="38" y="113">
    <field name="FIELDNAME">89</field>
  </block>
  <block type="test_fields_date" id="flqjVdY:q[i9i9quPeNH" x="37" y="163">
    <field name="FIELDNAME">2019-04-18</field>
  </block>
  <block type="test_fields_checkbox" id="ek{drzs2vnp^RIzb[~_." x="38" y="212">
    <field name="CHECKBOX">FALSE</field>
  </block>
  <block type="test_fields_colour" id="0kbOz)J7$}kH;aG$/7ku" x="38" y="263">
    <field name="COLOUR">#33ff33</field>
  </block>
  <block type="test_fields_text_input" id="f%yfBd#74I3,;do8vGAu" x="37" y="312">
    <field name="TEXT_INPUT">not default</field>
  </block>
  <block type="test_fields_variable" id="2![yICR}hguA1=0)-x`^" x="37" y="362">
    <field name="VARIABLE" id="o(EVLYWaZFZca|N:K.7I" variabletype="">not item</field>
  </block>
</xml>
```

Tested on:
* Desktop Chrome
<!-- * Desktop Firefox -->
<!-- * Desktop Safari -->
<!-- * Desktop Opera -->
<!-- * Windows Internet Explorer 10 -->
<!-- * Windows Internet Explorer 11 -->
<!-- * Windows Edge -->

<!--
* Smartphone/Tablet/Chromebook (please complete the following information):
  * Device: [e.g. iPhone6]
  * OS: [e.g. iOS8.1]
  * Browser [e.g. stock browser, safari]
  * Version [e.g. 22]
-->

### Additional Information

SERIALIZABLE is false by default. This is so that it is backwards compatible with any non-editable fields that people had created in the past. 
![IMG_20190418_082838](https://user-images.githubusercontent.com/25440652/56372575-196b9800-61b4-11e9-82da-3ff112748db1.jpg)
The problem with this is that if you had created a custom editable field before you'll get a warning message when it goes to serialize, because EDITABLE is true and SERIALIZABLE is not, which is an invalid state. It still serializes! But you will get that warning. I don't necessarily /like/ this but it's the best solution I could come up with.

Also I'm planning on adding field_label_serializable and field_image_serializable fields, but the field_label_serializable stuff started cluttering up this PR and the fromXML toXML change needs to get in before field_image_serializable can be added.

rachel-fenichel commented on core/field.js:
```diff
@@ -268,15 +282,34 @@ Blockly.Field.prototype.updateEditable = function() {
+ */
+Blockly.Field.prototype.isSerializable = function() {
+  var isSerializable = false;
+  if (this.name) {
```
Why does the name matter?

NeilFraser replied:
If there is no name, then serializing doesn't make sense since you can't refer to the field.

rachel-fenichel replied:
Oh right.

rachel-fenichel commented on core/field_angle.js:
```diff
@@ -66,6 +66,22 @@ Blockly.FieldAngle.fromJson = function(options) {
+ * @type {boolean}
+ * @const
+ */
+Blockly.FieldAngle.prototype.EDITABLE = true;
```
Since EDITABLE defaults to true on fields, you don't need to redeclare it on each of these classes.  You only need to declare it for classes that set it to false.  The inverse is true for SERIALIZABLE.

BeksOmega replied:
Fixed.

rachel-fenichel commented:
One lint error on travis: https://travis-ci.org/google/blockly/builds/521780332#L536

NeilFraser commented:
> One lint error on travis: https://travis-ci.org/google/blockly/builds/521780332#L536

That console log looks appropriate and useful.  What's Travis' logic that flags this console usage and not others?

BeksOmega commented:
> We don't have mocha running on our server yet--are you running these tests locally?

Yes I added them to your handy-dandy index.html file. Here's a pic of my added tests (ignore the variable fields tests):
![TestsPassing](https://user-images.githubusercontent.com/25440652/56396791-70de2800-61f5-11e9-8888-ff7f3122dfbd.jpg)

rachel-fenichel commented:
> That console log looks appropriate and useful. What's Travis' logic that flags this console usage and not others?

Not sure, actually.  it should all be running with the same eslint rules, since our eslintrc is in our root directory, and that has the no-console rule disabled.  Weird.

BeksOmega commented:
Ok @rachel-fenichel I changed the mocha eslint config to extend the root one, and then I fixed the lint errors in the other mocha files. I can separate those changes into a separate PR if you like, otherwise I think this is good to go.

rachel-fenichel commented:
Awesome, thanks for tracking that down.
